### ISSUE TITLE
Check that `B.__radd__` is a subtype of `A.__add__` for `B <: A`, and support `__radd__` priority.

### ISSUE THREAD

**Issue by randolf-scholz** [microsoft/pyright#9684]
Currently, `pyright` does not correctly handle the runtime behavior of operators, in particular the following runtime behavior is ignored: 

> If `B <: A`, Python tries `B.__radd__` before `A.__add__`. This is ok, because it was implemented with knowledge of `A`

See [[1]](https://docs.python.org/3/library/numbers.html#implementing-the-arithmetic-operations) and [[2]](https://docs.python.org/3/reference/datamodel.html#object.__ror__). This has been brought up before in #3862 and #3368, and was deemed unfixable due to:

> The problem is that when you specify a type annotation, you are not specifying an exact type. For example, if you specify that a parameter is of type `Foo`, it doesn't need to literally be an instance of the class `Foo`. It can be any subclass of `Foo`.  So if the type checker sees that there's a `+` operator and the LHS and RHS are both determined to be `Foo` or a subclass thereof, it cannot tell whether the runtime interpreter will apply this special-case rule and prioritize `__radd__` or `__add__`. 

 _Originally posted by @erictraut in [#3862](https://github.com/microsoft/pyright/issues/3862#issuecomment-1229425541)_

<details><summary> However, on the other hand, mypy actually implements this behavior, which causes divergence between the two type checkers.  </summary>

Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.385&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQFD0DGFZAzm1ACoCuCFhAFz0ooqABNCwKFSplx42QAo2hCsAA0UMDAAWhEIKgAiXv0LGAlFAC0APhNmBxo0Rg8QKbnwFLLjFnZOAGUkAC9CJSdCS2ExCSkZOQVlVXUtHX1Db3NrexNQiJcCQndPKELI-3jJaVkQeUUqFTVNbT0DI2i8h2NK4rcPL0q-ANYOKAAheWionxi4sQBiUQhEOCNQtBQyMsJOMGljesbZY20j6fFo87IUcRNZU6pzw8eqCDJUWQA6G6gyEQoDwUGwyMA1HBtKQQKwEMh0CIxLUkg0Us00m1Mp0cgIepgUDBXKUhlAAAyiFZrBAbYGg8GELQ0kBINC6IlQADyAGlGPQiKRKDREJFZtYANQVcJVaxQFYgEEwLBCKURLTU2mVJlwFlsjnRfkkcjUWiREYS3EyykERXKzbS9Xre1qqDM1ns50MIA),  [mypy playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=c0f5defa4649d5e97022689025d86c8c), 

```python
from typing import reveal_type

class Tuple:
    def __add__(self, other: "Tuple") -> "Tuple": return Tuple()

class Size(Tuple):
    def __add__(self, other: Tuple) -> "Size": return Size()
    def __radd__(self, other: Tuple) -> "Size": return Size()

class BadTuple(Tuple):
    #  mypy: Signatures of "__radd__" of "BadTuple" and "__add__" of "__main__.Tuple" are unsafely overlapping
    def __radd__(self, other: Tuple) -> int: return 0  # mypy: unsafe, pyright: OK


reveal_type(Tuple() + Size())  # runtime: Size, mypy: Size, pyright: Tuple
reveal_type(Size() + Tuple())  # runtime: Size, mypy: Size, pyright: Size
```

</details>

The above example shows that `mypy` also validates that `B.__radd__` is a subtype of `A.__add__` when defining the subclass, whereas `pyright` ignores this constraint. This is essential, because it means that if at runtime `B.__radd__` happens rather than `A.__add__`, we get a subtype of annotated return type, which by LSP should be unproblematic. This is just how subclasses behave generally when they are upcasted.

So it seems that simply adding this consistency constraint solves the problem outlined in https://github.com/microsoft/pyright/issues/3862#issuecomment-1229425541 and makes supporting the actual runtime behavior in a type-safe manner feasible.

<details> <summary> EDIT: full suite of all operators </summary>

```python
from typing import reveal_type

class A:
    # fmt: off
    def __add__(self, other: "A") -> "A": return A()
    def __sub__(self, other: "A") -> "A": return A()
    def __mul__(self, other: "A") -> "A": return A()
    def __truediv__(self, other: "A") -> "A": return A()
    def __floordiv__(self, other: "A") -> "A": return A()
    def __mod__(self, other: "A") -> "A": return A()
    def __divmod__(self, other: "A") -> "A": return A()
    def __pow__(self, other: "A") -> "A": return A()
    def __lshift__(self, other: "A") -> "A": return A()
    def __rshift__(self, other: "A") -> "A": return A()
    def __and__(self, other: "A") -> "A": return A()
    def __xor__(self, other: "A") -> "A": return A()
    def __or__(self, other: "A") -> "A": return A()
    def __matmul__(self, other: "A") -> "A": return A()
    # fmt: on


class B(A):
    # fmt: off
    # Normal (left) operations
    def __add__(self, other: A) -> "B": return B()
    def __sub__(self, other: A) -> "B": return B()
    def __mul__(self, other: A) -> "B": return B()
    def __truediv__(self, other: A) -> "B": return B()
    def __floordiv__(self, other: A) -> "B": return B()
    def __mod__(self, other: A) -> "B": return B()
    def __divmod__(self, other: A) -> "B": return B()
    def __pow__(self, other: A) -> "B": return B()
    def __lshift__(self, other: A) -> "B": return B()
    def __rshift__(self, other: A) -> "B": return B()
    def __and__(self, other: A) -> "B": return B()
    def __xor__(self, other: A) -> "B": return B()
    def __or__(self, other: A) -> "B": return B()
    def __matmul__(self, other: A) -> "B": return B()

    # Reverse operations
    def __radd__(self, other: A) -> "B": return B()
    def __rsub__(self, other: A) -> "B": return B()
    def __rmul__(self, other: A) -> "B": return B()
    def __rtruediv__(self, other: A) -> "B": return B()
    def __rfloordiv__(self, other: A) -> "B": return B()
    def __rmod__(self, other: A) -> "B": return B()
    def __rdivmod__(self, other: A) -> "B": return B()
    def __rpow__(self, other: A) -> "B": return B()
    def __rlshift__(self, other: A) -> "B": return B()
    def __rrshift__(self, other: A) -> "B": return B()
    def __rand__(self, other: A) -> "B": return B()
    def __rxor__(self, other: A) -> "B": return B()
    def __ror__(self, other: A) -> "B": return B()
    def __rmatmul__(self, other: A) -> "B": return B()
    # fmt: on

# arithmetic operations
reveal_type(A() + B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() + A())  # runtime: B, mypy: B, pyright: B
reveal_type(A() - B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() - A())  # runtime: B, mypy: B, pyright: B
reveal_type(A() * B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() * A())  # runtime: B, mypy: B, pyright: B
reveal_type(A() / B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() / A())  # runtime: B, mypy: B, pyright: B
reveal_type(A() // B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() // A())  # runtime: B, mypy: B, pyright: B

# boolean operations
reveal_type(A() & B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() & A())  # runtime: B, mypy: B, pyright: B
reveal_type(A() | B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() | A())  # runtime: B, mypy: B, pyright: B
reveal_type(A() ^ B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() ^ A())  # runtime: B, mypy: B, pyright: B

# shift operations
reveal_type(A() >> B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() >> A())  # runtime: B, mypy: B, pyright: B
reveal_type(A() << B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() << A())  # runtime: B, mypy: B, pyright: B

# modulo operations
reveal_type(A() % B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() % A())  # runtime: B, mypy: B, pyright: B
reveal_type(divmod(A(), B()))  # runtime: B, mypy: B, pyright: A
reveal_type(divmod(B(), A()))  # runtime: B, mypy: B, pyright: B

# matmul
reveal_type(A() @ B())  # runtime: B, mypy: B, pyright: A
reveal_type(B() @ A())  # runtime: B, mypy: B, pyright: B
```  

</details>

**Comment by erictraut** [comment]
This has been discussed before and rejected.

**Comment by randolf-scholz** [comment]
@erictraut What about at least checking that `B.__radd__` is a subtype of `A.__add__` when defining a subclass?

**Comment by randolf-scholz** [comment]
There are some important use cases for this as well, consider for example a vector class like

```python
class Vector:
    def __add__(self, other: "Vector") -> "Vector": ...
    def __radd__(self, other: "Vector") -> "Vector": ...
```

Now, let's say we wanted to create a subclass `MaskedVector`, that handles the case of missing values. Adding a `Vector` to a `MaskedVector` should logically result in a `MaskedVector` again, hence we want signatures like:


```python
class MaskedVector(Vector):
    def __add__(self, other: "Vector") -> "MaskedVector": ...
    def __radd__(self, other: "Vector") -> "MaskedVector": ...
```

But then we run into excactly this issue:

```python
reveal_type(Vector() + MaskedVector())  # runtime: MaskedVector, pyright: Vector
reveal_type(MaskedVector() + Vector())  # runtime: MaskedVector, pyright: MaskedVector
```

**Comment by jorenham** [comment]
This is also leading to incorrect behaviour in numpy with e.g. `numpy.float64 <: builtins.float`:

```py
reveal_type(f % f8) 
```

Pyright reports `float`, which is incorrect, and e.g. mypy correctly reports `numpy.float64`.
